File: leetcode/1022-sum-of-root-to-leaf-binary-numbers/SumOfRootToLeafBinaryNumbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int result = 0;
    
    private void helper(TreeNode root, int sum) {
        if (root == null) {
            return;
        }
        
        sum <<= 1;
        sum += root.val;
        
        if (root.left == null && root.right == null) {
            result += sum;
        } else {
            helper(root.left, sum);
            helper(root.right, sum);
        }
    }
    
    public int sumRootToLeaf(TreeNode root) {
        helper(root, 0);
        
        return result;
    }
}

File: leetcode/1022-sum-of-root-to-leaf-binary-numbers/sum-of-root-to-leaf-binary-numbers.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number}
 */
const sumRootToLeaf = function (root, sum = 0) {
  if (root == null) {
    return 0;
  }

  const newSum = (sum << 1) + root.val;

  if (!root.left && !root.right) {
    return newSum;
  }

  return sumRootToLeaf(root.left, newSum) + sumRootToLeaf(root.right, newSum);
};
